File: src/main/java/sds/offre/concept_metier/Formule.java
package sds.offre.concept_metier;

import java.util.Optional;

public class Formule {

    private IdFormule id;
    private Durée durée;
    private Prix prix;

    public Formule() {

    }

    public Optional<FormuleCreee> créé(IdFormule id, Prix prix, Durée durée) {
        if (estCréée()) return Optional.empty();

        this.id = id;
        this.prix = prix;
        this.durée = durée;
        return Optional.of(FormuleCreee.de(id, prix, durée));

    }

    public Optional<PrixFormuleChangee> changePrix(Prix nouveauPrix) {
        if (!estCréée()) return Optional.empty();
        if (prix.equals(nouveauPrix)) return Optional.empty();

        prix = nouveauPrix;
        return Optional.of(PrixFormuleChangee.de(id, prix));
    }

    private boolean estCréée() {
        return this.id != null;
    }

    public IdFormule id() {
        return id;
    }

    public Prix prixDeBase() {
        return prix;
    }

    public Durée durée() {
        return durée;
    }
}

File: src/main/java/sds/offre/concept_metier/Formule.kt
package sds.offre.concept_metier

import java.util.*

class Formule(
        val id: IdFormule,
        val durée: Durée,
        prix: Prix) {

    var prixDeBase = prix
        private set

    fun changePrix(nouveauPrix: Prix) {
        prixDeBase = nouveauPrix
    }

    fun lePrixEstIdentique(nouveauPrix: Prix): Boolean {
        return prixDeBase == nouveauPrix
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val formule = other as Formule?
        return id == formule!!.id
    }

    override fun hashCode(): Int {
        return Objects.hash(id)
    }

    override fun toString(): String {
        return "Formule{" +
                "id=" + id +
                ", durée=" + durée +
                ", prix=" + prixDeBase +
                '}'.toString()
    }
}
